File: CPPQEDscripts/tutorialCompositeRing.cc
// Copyright András Vukics 2006–2020. Distributed under the Boost Software License, Version 1.0. (See accompanying file LICENSE.txt)
#include "EvolutionComposite.h"

#include "ParticleTwoModes.h"

int main(int argc, char* argv[])
{
  ParameterTable p;

  evolution::Pars<> pe(p);
  particle::Pars pp(p);
  mode::ParsLossy       pmP(p,"P");
  mode::ParsPumpedLossy pmM(p,"M");
  particlecavity::ParsAlong ppcP(p,"P");
  particlecavity::ParsAlong ppcM(p,"M");

  ppcP.modeCav=MFT_PLUS; ppcM.modeCav=MFT_MINUS; 

  auto qmp=updateWithPicture(p,argc,argv,"--");

  particle::Ptr part (make(pp ,qmp));
  mode    ::Ptr plus (make(pmP,qmp));
  mode    ::Ptr minus(make(pmM,qmp));

  evolve<0>(wavePacket(pp)*init(pmP)*init(pmM),
            composite::make(
                            _<1,0>  (std::make_shared<ParticleAlongCavity>(plus ,part,ppcP)),
                            _<2,0>  (std::make_shared<ParticleAlongCavity>(minus,part,ppcM)),
                            _<1,2,0>(std::make_shared<ParticleTwoModes>(plus,minus,part,ppcP,ppcM))
                            ),
            pe);

  {
  const composite::result_of::Make<_<1,0>,_<2,0>,_<1,2,0> >::type
    system=composite::make(
                           _<1,0>  (std::make_shared<ParticleAlongCavity>(plus ,part,ppcP)),
                           _<2,0>  (std::make_shared<ParticleAlongCavity>(minus,part,ppcM)),
                           _<1,2,0>(std::make_shared<ParticleTwoModes>(plus,minus,part,ppcP,ppcM)));

  system->displayParameters(std::cout);
  }
  {
  const auto
    system=composite::make(
                           _<1,0>  (std::make_shared<ParticleAlongCavity>(plus ,part,ppcP)),
                           _<2,0>  (std::make_shared<ParticleAlongCavity>(minus,part,ppcM)),
                           _<1,2,0>(std::make_shared<ParticleTwoModes>(plus,minus,part,ppcP,ppcM)));

  system->displayParameters(std::cout);
  }

}

File: cpypyqed/scripts/tutorialCompositeRing.py
#!/usr/bin/env python3
# coding: utf-8
import sys
from cpypyqed import *

p=parameters.ParameterTable()

pe=evolution.Pars(p)

pp=particle.Pars(p)

pmP=mode.ParsLossy(p,"P")
pmM=mode.ParsLossy(p,"M")
ppcP=particlecavity.ParsAlong(p,"P")
ppcM=particlecavity.ParsAlong(p,"M")

ppcP.modeCav=ModeFunctionType.PLUS
ppcM.modeCav=ModeFunctionType.MINUS

qmp=updateWithPicture(p,sys.argv,"--")

part=particle.make(pp ,qmp)
plus =   mode.make(pmP,qmp)
minus=   mode.make(pmM,qmp)

evolve(particle.wavePacket(pp)**mode.init(pmP)**mode.init(pmM),
       makeComposite({(1,0):ParticleAlongCavity(plus ,part,ppcP),
                      (2,0):ParticleAlongCavity(minus,part,ppcM),
                      (1,2,0):ParticleTwoModes(plus,minus,part,ppcP,ppcM)}),
       pe)
